pkgs/development/interpreters/python/sitecustomize.py

```diff
@@ -21,9 +21,11 @@
 if paths:
     functools.reduce(lambda k, p: site.addsitedir(p, k), paths.split(':'), site._init_pathinfo())
 
-# Check whether we are in a venv. 
-# Note Python 2 does not support base_prefix so we assume we are not in a venv.
-in_venv = sys.version_info.major == 3 and sys.prefix != sys.base_prefix
+# Check whether we are in a venv or virtualenv. 
+# For Python 3 we check whether our `base_prefix` is different from our current `prefix`.
+# For Python 2 we check whether the non-standard `real_prefix` is set.
+# https://stackoverflow.com/questions/1871549/determine-if-python-is-running-inside-virtualenv
+in_venv = (sys.version_info.major == 3 and sys.prefix != sys.base_prefix) or (sys.version_info.major == 2 and hasattr(sys, "real_prefix"))
```

Comment:
pyvenv.cfg is still Python 3-only, which is a downside. An alternative I've also considered is setting instead `PYTHONUSERBASE`. But that won't be sufficient for the non-withPackages case such as executables in applications.

Anyway, there is no clear solution to this problem because the problem we need to solve is not a typical problem for which the tooling has been designed. It's a matter of balancing things to come to a "sufficiently working" setup.

At least now by improving test coverage we can test whether any new method will really be better or not. And this change is an improvement.

Reply:
> At least now by improving test coverage we can test whether any new method will really be better or not. And this change is an improvement.

Agree to that.

Right now i started to get a clue on the python nixos tests. 
I noticed we don't have integration tests for the cases we are in any virtual env, is this true or did i miss them?

I will try to extend that so that we can test 'import sqlite3' (for std lib) and 'import sitepkg' with any of the virtual env pythons.

Reply:
If we don't have them, do you think it is ok, to put right into the [testfun](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/tests.nix#L81) of  or should they live in a separate file?

Reply:
Yes, in `pkgs/development/interpreters/python/tests/test_python.py`. There it tests the prefixes for Nix envs, virtualenvs and venvs. That could be extended.

Note these are not NixOS tests, as in, they do not run in a VM.

Reply:
Yes, ok, then i will make a PR for that. Thanks, for the clarification also.

Reply:
@FRidh i have done it. But am waiting right now until my local staging is build to check if everything applies. Btw. as the change somehow depends on the merged commits to staging should this PR be opened against staging or master?

Reply:
against staging

Reply:
Ay, ok. that i never did before and was scared because of so many changes when github compared against master by default.